test/Microsoft.AspNetCore.SignalR.Tests/HubConnectionHandlerTests.cs

```diff
@@ -2332,6 +2333,202 @@ public int Serialize(ref byte[] bytes, int offset, T value, IFormatterResolver f
+
+            var serviceProvider = HubConnectionHandlerTestUtils.CreateServiceProvider();
+            var connectionHandler = serviceProvider.GetService<HubConnectionHandler<MethodHub>>();
+
+            using (var client = new TestClient())
+            {
+                var connectionHandlerTask = await client.ConnectAsync(connectionHandler).OrTimeout();
+                await client.BeginUploadStreamAsync("invocation", nameof(MethodHub.StreamingConcat), new StreamPlaceholder("id")).OrTimeout();
+
+                // send integers that are then cast to strings
+                await client.SendHubMessageAsync(new StreamDataMessage("id", 5)).OrTimeout();
+                await client.SendHubMessageAsync(new StreamDataMessage("id", 10)).OrTimeout();
+
+                await client.SendHubMessageAsync(new StreamCompleteMessage("id")).OrTimeout();
+                var response = (CompletionMessage)await client.ReadAsync().OrTimeout();
+                
+                Assert.Equal("510", response.Result);
+            }
+        }
+
+        [Fact]
+        public async Task UploadStreamItemInvalidType()
+        {
+            var serviceProvider = HubConnectionHandlerTestUtils.CreateServiceProvider();
+            var connectionHandler = serviceProvider.GetService<HubConnectionHandler<MethodHub>>();
+
+            using (var client = new TestClient())
+            {
+                var connectionHandlerTask = await client.ConnectAsync(connectionHandler).OrTimeout();
+                await client.BeginUploadStreamAsync("invocationId", nameof(MethodHub.TestTypeCastingErrors), new StreamPlaceholder("channelId")).OrTimeout();
+
+                // client is running wild, sending strings not ints. 
+                // this error should be propogated to the user's HubMethod code
+                await client.SendHubMessageAsync(new StreamItemMessage("channelId", "not a number")).OrTimeout();
+                var response = await client.ReadAsync().OrTimeout();
+
+                Assert.Equal(typeof(CompletionMessage), response.GetType());
+                Assert.Equal("error identified and caught", (string)((CompletionMessage)response).Result);
+            }
+        }
+
+        [Fact]
+        public async Task UploadStreamItemInvalidId()
+        {
+            var serviceProvider = HubConnectionHandlerTestUtils.CreateServiceProvider();
+            var connectionHandler = serviceProvider.GetService<HubConnectionHandler<MethodHub>>();
+
+            using (var client = new TestClient())
+            {
+                var connectionHandlerTask = await client.ConnectAsync(connectionHandler).OrTimeout();
+                await client.SendHubMessageAsync(new StreamItemMessage("fake_id", "not a number")).OrTimeout();
+
+                // Client is breaking protocol by sending an invalid id, and should be closed.
+                var message = client.TryRead();
+                Assert.True(message is CloseMessage);
```

Comment:
two things:

1. Use `Assert.IsType<CloseMessage>(message)`, it has a cleaner error message, **and** it returns the object casted to `CloseMessage` for you to use in later Asserts
2. We should assert the error message as well.

Reply:
Current error message is the generic "Connection closed with an error".

Want to enable the detailed error messages for better testing, but not sure how to toggle the option through DI stuff.